helix-parsec/src/lib.rs

```diff
@@ -0,0 +1,580 @@
+// parser(s) as input and return a new parser: a function that takes input and returns
+// a parse result. The underlying implementation of [Parser::parse] for these functions
+// is simply application.
+#[doc(hidden)]
+impl<'a, F, T> Parser<'a> for F
+where
+    F: Fn(&'a str) -> ParseResult<T>,
+{
+    type Output = T;
+
+    fn parse(&self, input: &'a str) -> ParseResult<'a, Self::Output> {
+        self(input)
+    }
+}
+
+/// A parser which matches the string literal exactly.
+///
+/// This parser succeeds if the next characters in the input are equal to the given
+/// string literal.
+///
+/// Note that [str::parse] interferes with calling [Parser::parse] on string literals
+/// directly; this trait implementation works when used within any parser combinator
+/// but does not work on its own. To call [Parser::parse] on a parser for a string
+/// literal, use the [token] parser.
+///
+/// # Examples
+///
+/// ```
+/// use helix_parsec::{or, Parser};
+/// let parser = or("foo", "bar");
+/// assert_eq!(Ok(("", "foo")), parser.parse("foo"));
+/// assert_eq!(Ok(("", "bar")), parser.parse("bar"));
+/// assert_eq!(Err("baz"), parser.parse("baz"));
+/// ```
+impl<'a> Parser<'a> for &'static str {
+    type Output = &'a str;
+
+    fn parse(&self, input: &'a str) -> ParseResult<'a, Self::Output> {
+        match input.get(0..self.len()) {
+            Some(actual) if actual == *self => Ok((&input[self.len()..], &input[0..self.len()])),
+            _ => Err(input),
+        }
+    }
+}
+
+// Parsers
+
+/// A parser which matches the given string literally.
+///
+/// This function is a convenience for interpreting string literals as parsers
+/// and is only necessary to avoid conflict with [str::parse]. See the documentation
+/// for the `&'static str` implementation of [Parser].
+///
+/// # Examples
+///
+/// ```
+/// use helix_parsec::{token, Parser};
+/// let parser = token("foo");
+/// assert_eq!(Ok(("", "foo")), parser.parse("foo"));
+/// assert_eq!(Err("bar"), parser.parse("bar"));
+/// ```
+pub fn token<'a>(literal: &'static str) -> impl Parser<'a, Output = &'a str> {
+    literal
+}
+
+/// A parser which matches the pattern described by the given regular expression.
+///
+/// The pattern must match from the beginning of the input as if the regular expression
+/// included the `^` anchor. Using a `^` anchor in the regular expression is
+/// recommended in order to reduce any work done by the regex on non-matching input.
+///
+/// # Examples
+///
+/// ```
+/// use helix_parsec::{pattern, Parser};
+/// use regex::Regex;
+/// let regex = Regex::new(r"Hello, \w+!").unwrap();
+/// let parser = pattern(&regex);
+/// assert_eq!(Ok(("", "Hello, world!")), parser.parse("Hello, world!"));
+/// assert_eq!(Err("Hey, you!"), parser.parse("Hey, you!"));
+/// assert_eq!(Err("Oh Hello, world!"), parser.parse("Oh Hello, world!"));
+/// ```
+pub fn pattern<'a>(regex: &'a Regex) -> impl Parser<'a, Output = &'a str> {
```

Comment:
We could make it a feature flag?

Reply:
I wish we could use the generic https://doc.rust-lang.org/std/str/pattern/trait.Pattern.html but it's still not stable

Reply:
I think if one of our combinators is not good enough to match a certain pattern then we should just add a new one. That's sort of the nice thing about parser combinators (and keeping the crate in-tree). After having to write a bunch of parsers I have come to the conclusions that handrolled code is always preferable to a regex when writing a "hardcoded" (not adjusted at runtime) parser like this. It's faster at runtime, has less compile-time overhead, easier to debug, easier to understand, easier extend and can produce better errors. 

`regex` is great but more for runtime pattern matching rather than handwritten parsers/lexers.

> I wish we could use the generic [doc.rust-lang.org/std/str/pattern/trait.Pattern.html](https://doc.rust-lang.org/std/str/pattern/trait.Pattern.html) but it's still not stable

I do wish that was stabilized tough but nothing seems to be happening in that regard :/